executor/adapter.go

```diff
@@ -720,16 +724,25 @@ func (a *ExecStmt) handleNoDelayExecutor(ctx context.Context, e Executor) (sqlex
 	return nil, err
 }
 
-func (a *ExecStmt) handlePessimisticDML(ctx context.Context, e Executor) error {
+func (a *ExecStmt) handlePessimisticDML(ctx context.Context, e Executor) (err error) {
 	sctx := a.Ctx
-	// Do not active the transaction here.
+	// Do not activate the transaction here.
 	// When autocommit = 0 and transaction in pessimistic mode,
 	// statements like set xxx = xxx; should not active the transaction.
 	txn, err := sctx.Txn(false)
 	if err != nil {
 		return err
 	}
 	txnCtx := sctx.GetSessionVars().TxnCtx
+	defer func() {
+		// If it's not a retryable error, rollback current transaction instead of rolling back current statement like
+		// in normal transactions, because we cannot locate and rollback the statement that leads to the lock error.
+		// This is too strict, but since the feature is not for everyone, it's the easiest way to guarantee safety.
+		if err != nil && !sctx.GetSessionVars().ConstraintCheckInPlacePessimistic {
+			sctx.GetSessionVars().SetInTxn(false)
```

Comment:
Maybe an `INFO` log could be added here to record the error information and execution details, users may not be able to resolve this error by themselves and there could be oncall issues.